feature-libs/cart/base/components/cart-shared/cart-item-list/cart-item-list.component.ts

```diff
@@ -61,6 +61,7 @@ export class CartItemListComponent implements OnInit, OnDestroy {
   @Input() cartId: string;
 
   protected _items: OrderEntry[] = [];
+  protected _forceRerender: boolean = false;
```

Comment:
[MEDIUM]
Introducing this `_forceRerender` property which couples methods `getInputsFromContext()` with `rerenderChangedItems()` makes it much harder for me to understand how the component works.
I guess, it was introduced to workaround to some issue with change detection.

What other alternatives did we consider to fix the issue. What pros and cons of them do you see?

Reply:
@Platonn As far as us remember the issue was the following. As you can see in the type script doc of the `rerenderChangedItem` method, the component only re-renders any cart item when it was actually changed in an attempt to optimize performance

> We'd like to avoid the unnecessary re-renders of unchanged cart items after the data reload. OCC cart entries don't have any unique identifier that we could use in Angular `trackBy`.So we update each array element to the new object only when it's any different to the previous one.

For quote there is one case where this is not sufficient. In case a quote is switched from read-only to edit mode or vice versa, a re-render is also required, as an item is rendered differently in read-only mode compared to edit mode. For example in edit mode there is a stepper to enter a quantity, while the same control is disabled in read-only mode.

So when switching the quote from read-only to edit mode, the state of the read-only flag in the context will change and the component shall re-render.

The way its implemented is in my eyes the straight forward way of fixing the issue. In case the read-only flag from context switches, we need to re-render any item once, no matter whether the item was changed or not.

But maybe i'm missing here something, so let me know if you have any other idea, better way of fixing it.

Reply:
I agree your solution is probably the least invasive one that could exist. It's because the existing TS of this component is already complex.

Here's how I understand the reasons of the existing complexity:
1. Input data to this component can be passed in 2 possible ways:
- via regular `@Input()`s passed by a parent component
- via injectable and observable Context
The values from the injectable Context are copied to the regular properties being `@Input()`s

2. Setting new value for some`@Input()` properties triggers a complex logic, e.g. the setter of the proptery `items` triggers 2 functions: `resolveItems(items)` and `createForm()` (which depend only on the value of `items`). 
3. In some cases, **the logic** triggered by a change of some input properties **depends also on values of other properties/inputs**, which is prone to race conditions, when the properties are changed in a different order. For example: the setter of the proeprty `isCartLodaing` depends on the value of `readonly` ([see source](https://github.com/SAP/spartacus/blob/1f14dfde266d880c90cfe65740476f29367262da/feature-libs/cart/base/components/cart-shared/cart-item-list/cart-item-list.component.ts#L78-L87)) - please note that in this case the order updating values of properties `isCartLoading` and `readonly` **matter**. In particular, if the logic for the setter of `isCartLoading` may use the outdated value of `readonly`, if the property `readonly` is updated only after the `isCartLoading`.

This PR adds up to the complexity, by introducing the property `_forceRerender` which is set on change of the  property `readonly`, but used (and reset) in the setter's logic of the property `items`:
1) In this PR, we propose that the function `resolveItems(items)` triggered by the setter of the property `items` now depends on the value of `readonly` (or to be precise: on the fact whether the property `readonly` just have changed). This is prone to a race condition issues, if the property `readonly` is updated only after the property `items` is updated - because then the logic of `resolveItems(items)` won't know yet about the latest value (and also about the change) of the property `readonly`.

2) Moreover this PR proposes to save an information about a change of the property `readonly` to the boolean property `_forceRerender` only, when the value of `readonly` is changed **only by the Context.** But to be 100% correct, we should behave similarly, also if the property `readonly` is changed via a regular `@Input()` - this is not handled now, as far as I understand from the code.

3) The information about the change of the property `readonly` is stored in the class state, in a property `_forceRerender`. Then this property is reset to false by some other method. Ideally the such a volatile information as "property has changed" should be passed directly as an argument to a logic that needs this information at the moment of the change. Storing it in a property and then resetting in some other method makes the class logic harder to reason about.

4) Now we call `cdr.markForCheck()` only after the change of the property `readonly`, but not in the case of chaning other properties, which is not obvious _why_, on the first sight. Btw.  _[QUESTION]: Do we really need to call `markForCheck()` when property `readonly` changes? If yes, don't we need to do it for the update of any other property?_

I can understand that because of the existing complexity of this class, it's challenging to find a simple solution to a bug (bug is the following, as far as I understood: the form controls are not updated in the html view when the property `readonly` changes its value).

To avoid having race conditions in terms of setting various input properties, we should avoid triggering any setters logic on individual input properties, but instead trigger all such logic **only after all inputs are updated**. In particular, we should do it in those moments:
- when input properties are changed by the observable Context -> in the end of the subscription's callback inside `getInputsFromContext()`
- when input properties are updated via regular `@Input()`s passed by a parent component - in `ngOnChanges()` lifecycle hook
Here's a snippet of code illustrating this idea:

```ts
export class CartItemListComponent implements OnInit, OnChanges, OnDestroy {
  /* ... */
 @Input('items')
  set items(items: OrderEntry[]) {} // logic moved to `ngOnChanges`
  get items(): OrderEntry[] {
    return this._items;
  }
  
  /* ... */
  
  @Input('cartIsLoading') set setLoading(value: boolean) {} // logic moved to `ngOnChanges`
  
  /* ... */

  protected getInputsFromContext(): Subscription | undefined {
    return this.outlet?.context$.subscribe((context) => {
      let shouldMarkForCheck = false;
      if (context.readonly !== undefined) {
        this.readonly = context.readonly;
      }
      if (context.hasHeader !== undefined) {
        this.hasHeader = context.hasHeader;
      }
      if (context.options !== undefined) {
        this.options = context.options;
      }
      if (context.cartId !== undefined) {
        this.cartId = context.cartId;
      }
      if (context.items !== undefined) {
        this.items = context.items;
      }
      if (context.promotionLocation !== undefined) {
        this.promotionLocation = context.promotionLocation;
      }
      if (context.cartIsLoading !== undefined) {
        this.setLoading = context.cartIsLoading;
      }

      // trigger calling `ngOnChanges` manually:
      this.ngOnChanges(/* SimpleChanges object created from the Context */);
    });
  }

  // triggered on regular `@Input()`s change OR when outlet context changes:
  ngOnChanges(changes: SimpleChanges): void {
    const readonlyChanged = changes.readonly?.currentValue !== changes.readonly?.previousValue;
    const itemsChanged = changes.items?.currentValue !== changes.items?.previousValue;

    if(itemsChanged || readonlyChanged) {
      this.resolveItems(changes.items.currentValue, {resetAllItems: readonlyChanged});
      this.createForm();
    }

    // ---

    const cartIsLoadingChanged = changes.cartIsLoading?.currentValue !== changes.cartIsLoading?.previousValue;
    if(cartIsLoadingChanged && !this.readonly) {
      // Whenever the cart is loading, we disable the complete form
      // to avoid any user interaction with the cart.
      this.cartIsLoading
        ? this.form.disable({ emitEvent: false })
        : this.form.enable({ emitEvent: false });
      this.cd.markForCheck();
    }
  }
 
  // please note new optional argument `{ resetAllItems: boolean }`, which is true,
  // when the property `readonly` has changed, and can be used later in the method `rerenderChangedItems()
  // also via an optional new argument
  protected resolveItems(items: OrderEntry[], options?: {resetAllItems?: boolean}): void {
    / * ... */
      this.rerenderChangedItems(items, { resetAllItems: options?.resetAllitems } );
   }
   
    / * ... */
  
  // `resetAllItems` is used eventually in this method
  protected rerenderChangedItems(items: OrderEntry[], options?: {resetAllItems?: boolean} ) {
    / * ... */
    if (
        options.resetAllItems ||
        JSON.stringify(this._items?.[index]) !== JSON.stringify(items[index])
      ) {
    /* ... */
   }
```

What do you think about it?

Reply:
I have not seen this component before in detail, therefore hopefully I am not missing anything:
I understand your concerns Kris, agree that the component is too complex already now (and maybe depends on some assumptions that are not always met), and that the quote extension even increases the concerns. What you propose sounds plausible.
However I see 2 risks for the refactoring: (1) Applying the feature flags for covering the breaking changes will again increase the complexity and too is error-prone
(2) We are probably not aware of all possible cart features and usages of this components, so we could create regressions.

In case we do the the proposed refactoring, we also need to involve someone who has a bit of knowledge about all SPA cart features..(or would that be you Kris?). That would at least mitigate the second risk. What do you think @Uli-Tiger and @Platonn ?

Reply:
> [QUESTION]: Do we really need to call markForCheck() when property readonly changes? If yes, don't we need to do it for the update of any other property?
If i remove the markForCheck() the issue occurs again. I tested  a bit and it seems that it needs to be called before the items setter is called, which in turn triggers the re-render logic.

Reply:
I refactored the code a bit to address at least some concerns, but still keeping the amount of changes limited, so that i have confidence that my changes do not break anything, especially considering that i do not know all the use case of the component. Especially the _forceReRender property was eliminated, instead passing on the flag as optional method parameter.
The fact that i do not know all the use cases of the component, and that i do not know why it was written in this way in the first place is for strong indicator to avoid a full refactoring potentially breaking scenarios i do not know or understand.